File: meo_so/week01/MTA/js/MTA.js
console.log("This is MTA");

//create object holding all train lines in array
const trainLines = {
  N: ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'],
  L: ['8th', '6th', 'Union Square', '3rd', '1st'],
  '6': ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place']
};

// function to define train lines 
// input: train line name (string)
// output: clone of train line (array)
const getLine = function (line) {
    return Array.from(trainLines[line]);
};

// function to get segment of train line 
// input: train line (array), starting station name (string), destination station on same line (string)
// output: stations passed on the train line (array)
const lineSeg = function (line, start, finish) {
    const endIndex = line.indexOf(finish);
    const startIndex = line.indexOf(start);
    // using station index to calculate the direction of the ride
    if (startIndex < endIndex) {
        return line.splice(startIndex+1, (endIndex - startIndex));
    };
    return line.splice(endIndex, (startIndex - endIndex)).reverse();
};

const tripMessage1 = function (line, stationsPassed) {
    return `You must travel through the following stops on the ${ line } line: ${ stationsPassed.join(', ') }.`;
}

const checkInput = function (startLine, startStation, finalLine, finalStation) {
    const errorMessage = [];
    if (!Object.keys(trainLines).includes(startLine) || !Object.keys(trainLines).includes(finalLine)) {
        errorMessage.push('Invalid lines');
    } else if (getLine(startLine).indexOf(startStation) < 0 || getLine(finalLine).indexOf(finalStation) < 0) {
        errorMessage.push('Invalid station names');
    };
    return errorMessage;
};

// function to run planTrip
// input: first train line name (string), firt station name (string), destination train line (string), 
// destination station name (string).
// first check the input train line and station are valid. 
// output: error message if input are invalid. exit function. 
// then check if the trip involve only one train line, 
// output: print details of one train line trip and total stops. exit function. 
// if trips involve two train line: 
// output: print first trip details then print second trip details. Print total stops involed. exit function.
const planTrip = function (startLine, startStation, finalLine, finalStation) {
    // input checking
    const errorMessage = checkInput(startLine, startStation, finalLine, finalStation);
    if (errorMessage.length > 0) {
        return console.log(errorMessage);
    // when trip happen on one line only
    } else if (startLine === finalLine) {
        const stopsPassed  = lineSeg(getLine(startLine), startStation, finalStation);
        return console.log(tripMessage1(startLine, stopsPassed) + `\n${ stopsPassed.length} stops in total.`);
    };
    // stations travelled on starting line
    const firstTrip = lineSeg(getLine(startLine), startStation, 'Union Square'); //trip from start to US station
    // only show trip on line when array is not empty
    console.log((firstTrip.length > 0) ? (tripMessage1(startLine, firstTrip) + `\nChange at Union Square.`) : `\nChange at Union Square.`);
    // stations travelled on finish line
    const secTrip = lineSeg(getLine(finalLine), 'Union Square', finalStation); //trip from US to final stop.
        console.log(tripMessage1(finalLine, secTrip) + `\n${ (firstTrip.length + secTrip.length) } stops in total.`);
    };

let run = function () {
    let promprom= true;
    if (promprom){
        planTrip(prompt(`starting train line.`), prompt('starting station'), prompt('destination train line'), prompt('destination station'));
    }else{
        console.log('wrong line name');
        planTrip('LD', '23rd', 'N', '8th');
        console.log('\none trip on N 23rd to N 8th');
        planTrip('N', '23rd', 'N', '8th');
        console.log('\nwrong name');
        planTrip('6', '35th', 'N', 'Union Square'); //wrong name
        console.log('\nfrom 6line Grand Central one end to another , Astor Place');
        planTrip('6', 'Grand Central', '6', 'Astor Place'); // from 6line one end to another
        console.log('\nfrom N 8th to L 1st');
        planTrip('N', '8th', 'L', '1st'); // from N to L
        console.log('\nlongest from N Times Square to 6 Grand Central');
        planTrip('N', 'Times Square', '6', 'Grand Central'); //'longest from N to 6'
        console.log('\nstart from N union square to 6 Astor Place');
        planTrip('N', 'Union Square', '6', 'Astor Place');
                
    }
}
run();

File: meo_so/week04/MTA/MTA.rb
require 'pry'
TRAIN_LINES = { 
    :N => ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
    :L => ["8th", "6th", "Union Square", "3rd", "1st"],
    :"6" => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]
}
# get array of single trip on a line
def one_trip (start_line, start_station, end_station)
    start_index = TRAIN_LINES[start_line].index(start_station)
    end_index =  TRAIN_LINES[start_line].index(end_station)
    return end_index > start_index ? TRAIN_LINES[start_line][start_index+1..end_index] : TRAIN_LINES[start_line][end_index...start_index].reverse
end

#calculate trip details
def plan_trip (start_line, start_station, end_line, end_station)
    trainList = TRAIN_LINES.keys
    if !trainList.include?(start_line) || !trainList.include?(end_line)   # check train line
        puts "Wrong train line, please enter valid train line"
    elsif !TRAIN_LINES[start_line].include?(start_station) || !TRAIN_LINES[end_line].include?(end_station)  #check station name
        puts "Wrong station name or does not exist"
    elsif start_line == end_line   # case of only travel 1 line
        one_line_trip = one_trip start_line, start_station, end_station
        puts "You must travel through the following stops on the N line: #{ one_line_trip.join(', ') }."
        puts "#{one_line_trip.size} stops in total."
    else   #full trip
        first_trip = one_trip start_line, start_station, "Union Square"
        sec_trip = one_trip end_line, "Union Square", end_station
        puts "You Must travel through the following stops on the #{start_line} line: #{first_trip.join(', ')}."
        puts "Change at Union Square."
        puts "Your jorney continues through the following stops: #{sec_trip.join(', ')}."
        puts "#{(first_trip.size + sec_trip.size)} stops in total."
    end 
end

# to get data from user
def input_trip
    puts "Please enter starting train line "
    start_line = gets.chomp.to_s.capitalize.to_sym
    puts "Please enter starting station "
    start_station = gets.chomp.to_s.split.map(&:capitalize)*' '
    puts "Please enter finish train line "
    end_line = gets.chomp.to_s.capitalize.to_sym
    puts "Please eneter finish station "
    end_station = gets.chomp.to_s.split.map(&:capitalize)*' '
    plan_trip start_line, start_station, end_line, end_station
end
input_trip